cEP-0021.md

```diff
@@ -0,0 +1,164 @@
+  `.coafile` i.e. `profile_bears = [Dirname]`
+  It will dump the *raw profile output* on the specified
+  directory, i.e., [Dirname]
+
+The profiler will start by profiling the `run()` method of the bears
+The advantage of profiling just the `run()` method is that bear writer
+don't have to spend time on loading the files, collecting the settings,
+etc. as this is all done by coala itself. They just have to profile
+the `run()` method to create a Better Performant bear.
+
+# Implementation Details
+
+When user invokes the profiler with either *CLI* or by `coafile`
+then `execute` method of Bear base class will call the `run_bear_from_section`
+method.
+
+```python
+def profile(self, dirnames = ['.']):
+  ...
+
+def run_bear_from_section(self, args, kwargs):
+  ...
+  profile_bears = kwargs.get('profile_bears')
+  if profile_bears is None:
+      self.run(*args, **kwargs)
+  elif:
+      self.profile(*args, **kwargs)
+  ...
+```
+
+`run_bear_from_section` method checks whether bear writer invoke the profiler
+interface or not i.e., keyword arguments passed to the Bear contain the
+`profile_bears` key or not.
+
+- If *not* then `run_bear_from_section()` method will directly call
+  the `run()` method of Bear so that coala can continue its normal
+  execution.
+- Else `run_bear_from_section()` method will call `profile()` method which
+  will check the whether the value of `profile_bears` key value is
+  `True` or `[Dirname]`.
+  - If `True` then `profile()` will create a profiler file for every
+    bear of every section with a name `{section_name}{bear_name}.prof` and
+    dump it in a *current working directory.*
+
+```python
+kwargs.pop('profile_bears')
+path =  os.getcwd()
+if str(profile_dump)=='True':
+    filename = '{}_{}.prof'.format(
+        kwargs['section_name'], kwargs['bear_name'])
+    open(filename, 'w+')
+    kwargs.pop('section_name')
+    kwargs.pop('bear_name')
+    cProfile.runctx('self.run(*args, **kwargs)',
+                    globals(), locals(), filename)
+```
+
+- If `[Dirname]` then `profile()` method will create a profiler file for
+  every bear of every section with a name `{section_name}{bear_name}.prof` and
+  after that it check whether the specified directory exits or not. <br>
+  If exits then it will directly dump the profiler files on the specified
+  directory <br> If not then it will create a subdirectory name `[Dirname]`
+  and dump the profiler files on that.
+
+```python
+subdir = str(profile_dump)
+filepath = os.path.join(path, subdir)
+if os.path.isdir(filepath):
+    os.chdir(subdir)
+else:
+    os.mkdir(os.path.join(path, subdir))
+    os.chdir(subdir)
+filename = '{}_{}.prof'.format(
+    kwargs['section_name'], kwargs['bear_name'])
+open(filename, 'w+')
+kwargs.pop('section_name')
+kwargs.pop('bear_name')
+cProfile.runctx('self.run(*args, **kwargs)',
+                globals(), locals(), filename)
+os.chdir("..")
+```
+
+These profiler files contains a binary output which can be used for further
```

Comment:
Please also add a `##` Heading before this paragraph called 'Profiler Output' or similar.

Then start this Paragraph with:

The profiler produces binary output that can be displayed and formatted by third party tools.